http/src/main/scala/org/http4s/blaze/http/http2/Http2FrameDecoder.scala

```diff
@@ -0,0 +1,310 @@
+      buffer.reset()
+      BufferUnderflow
+    } else { // full frame. Decode.
+      // set frame sizes in the ByteBuffer and decode
+      val oldLimit = buffer.limit
+      val endOfFrame = buffer.position + len
+      buffer.limit(endOfFrame)
+
+      try frameType match {
+        case FrameTypes.DATA          => decodeDataFrame(buffer, streamId, flags)
+        case FrameTypes.HEADERS       => decodeHeaderFrame(buffer, streamId, flags)
+        case FrameTypes.PRIORITY      => decodePriorityFrame(buffer, streamId, flags)
+        case FrameTypes.RST_STREAM    => decodeRstStreamFrame(buffer, streamId)
+        case FrameTypes.SETTINGS      => decodeSettingsFrame(buffer, streamId, flags)
+        case FrameTypes.PUSH_PROMISE  => decodePushPromiseFrame(buffer, streamId, flags)
+        case FrameTypes.PING          => decodePingFrame(buffer, streamId, flags)
+        case FrameTypes.GOAWAY        => decodeGoAwayFrame(buffer, streamId)
+        case FrameTypes.WINDOW_UPDATE => decodeWindowUpdateFrame(buffer, streamId)
+        case FrameTypes.CONTINUATION  => decodeContinuationFrame(buffer, streamId, flags)
+
+        // this concludes the types established by HTTP/2.0, but it could be an extension
+        case code                     => onExtensionFrame(code, streamId, flags, buffer.slice())
+      } catch {
+        case _: BufferUnderflowException =>
+          Error(FRAME_SIZE_ERROR.goaway
+          (s"Frame type 0x${Integer.toHexString(frameType.toInt)} and size $len underflowed"))
+      } finally {
+        // reset buffer limits
+        buffer.limit(oldLimit)
+        buffer.position(endOfFrame)
+        ()
+      }
+    }
+  }
+
+  /** Hook method for handling protocol extension
+    *
+    * https://tools.ietf.org/html/rfc7540#section-5.5
+    *
+    * @param code the frame type of this extension frame
+    * @param streamId stream id associated with this extension frame
+    * @param flags the flags associated with this extension frame
+    * @param buffer the payload associated with this extension frame
+    * @return result of handling the message. If this extension frame is of
+    *         unknown type, it MUST be ignored by spec.
+    */
+  def onExtensionFrame(code: Byte, streamId: Int, flags: Byte, buffer: ByteBuffer): Http2Result =
+    Continue
+
+  //////////////// Decoding algorithms ///////////////////////////////////////////////////////////
+
+  //////////// DATA ///////////////
+  // https://tools.ietf.org/html/rfc7540#section-6.1
+  private def decodeDataFrame(buffer: ByteBuffer, streamId: Int, flags: Byte): Http2Result = {
+    if (streamId == 0) {
+      /*
+       DATA frames MUST be associated with a stream.  If a DATA frame is
+       received whose stream identifier field is 0x0, the recipient MUST
+       respond with a connection error (Section 5.4.1) of type
+       PROTOCOL_ERROR.
+       */
+      return Error(PROTOCOL_ERROR.goaway("Data frame with streamID 0x0"))
```

Comment:
I know we're where the grickle-grass grows, but do we really need `return` statements?

Reply:
Some of these get really nested otherwise, but if you feel strongly about it I can change it. Other than one at the top I believe all `return`s are errors.